src/Microsoft.VisualStudio.Sdk.TestFramework/Mocks/MockProfferAsyncService.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#if NETFRAMEWORK || WINDOWS
+
+using System.IO;
+using Microsoft;
+using Microsoft.VisualStudio.Shell.Interop;
+using ComAsyncServiceProvider = Microsoft.VisualStudio.Shell.Interop.COMAsyncServiceProvider.IAsyncServiceProvider;
+
+namespace Microsoft.VisualStudio.Sdk.TestFramework.Mocks;
+
+/// <summary>
+/// A working implementation of <see cref="IProfferAsyncService"/>.
+/// </summary>
+/// <param name="globalServiceProvider">The global service provider to which proffered services will be added.</param>
+/// <param name="logger">A means to log progress as it is reported.</param>
+public class MockProfferAsyncService(GlobalServiceProvider? globalServiceProvider, Action<string>? logger) : IProfferAsyncService
+{
+    private readonly List<Guid?> profferedCookies = new();
+
+    /// <summary>
+    /// Gets a dictionary of proffered services.
+    /// </summary>
+    public Dictionary<Guid, ComAsyncServiceProvider> ProfferedServices { get; } = new();
+
+    /// <summary>
+    /// Checks whether a service has been proffered.
+    /// </summary>
+    /// <param name="serviceType">The <see cref="Type"/> used for the service IID.</param>
+    /// <returns><see langword="true"/> if the service is proffered; otherwise <see langword="false" />.</returns>
+    public bool IsServiceProffered(Type serviceType)
+    {
+        Requires.NotNull(serviceType);
+        return this.ProfferedServices.ContainsKey(serviceType.GUID);
+    }
+
+    /// <inheritdoc/>
+    public IAsyncProgressCallback GetServiceProgressCallback()
+    {
+        return new MockAsyncProgressCallback(logger);
+    }
+
+    /// <inheritdoc/>
+    uint IProfferAsyncService.ProfferAsyncService(ref Guid rguidService, ComAsyncServiceProvider psp)
```

Comment:
the semantics of the shell proffer service are if you register an already registered service, it returns a cookie of 0 and S_FALSE, and a revoke on cookie 0 is just a NOP returning S_OK

Reply:
There are also service proffering events raised after registration / revocation (IServiceProfferingEventSink), but those are in the internal interop so not sure if you want to replicate that here

Reply:
The proffering events probably aren't interesting (and if they are, I can add them later). 
I've already documented behavior that calling AddService will overwrite a prior proffer, in order that built-in mocks can be replaced by a test that has special requirements. 
But clearly the cookie I return should not allow 0 for the first one. I'll fix that.